Common/include/toolboxes/CNewtonUpdateOnSubspace.hpp

```diff
@@ -0,0 +1,306 @@
+/*!
+ * \file CNewtonUpdateOnSubspace.hpp
+ * \brief
+ * \note Adopts some general layout from CQuasiNewtonInvLeastSquares.
+ * \author O. Burghardt
+ * \version 7.1.1 "Blackbird"
+ *
+ * SU2 Project Website: https://su2code.github.io
+ *
+ * The SU2 Project is maintained by the SU2 Foundation
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2020, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "CQuasiNewtonInvLeastSquares.hpp"
+
+#include <Eigen/Core>
+#include <Eigen/Dense>
+#include <Eigen/QR>
+//typedef Eigen::Matrix<su2double, Eigen::Dynamic, Eigen::Dynamic> EigenMatrix;
+//typedef Eigen::Matrix<su2double, Eigen::Dynamic, 1> EigenVector;
+//#include <Eigen/LU>
+//#include <Eigen/SVD>
+//typedef Eigen::JacobiSVD<EigenMatrix> EigenSVD;
+
+/*--- Both classes operate on a window of past corrected solutions (X) and an
+ *    data structure (R) of similar size to construct a (quasi-) Newton scheme.
+ *    Have to identify a proper common base class. ---*/
+template<class Scalar_t>
+class CNewtonUpdateOnSubspace : public CQuasiNewtonInvLeastSquares<Scalar_t> {
```

Comment:
Is Stroff the reference for this? Or are there modifications to the original RPM?

Reply:
It is very close to the original paper by Shroff. They didn't have access to AD derivatives of course, but apart from that, it's an implementation of the RPM algorithm proposed there (forgot to mention "RPM" above, thanks for the hint).

As a next step, I'd like to try out how keeping some "slow" basis vectors across outer loops, so that the whole multiphysics thing will hopefully have any gain from them. (Doing it for inner iterations won't help as much, as it takes some iterations before a proper subspace can be formed).

Reply:
:+1: I recently found out I should have named the other one "Anderson Acceleration"... 1965 such a hipster method.

Reply:
:O Great. There's even a Wiki article on that.. the chapter on relations to other classes of methods is quite interesting, too. Potentially this becomes a corrector toolbox family soon :)